// key-node/build/KeyNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _a;
"use strict";
const KeyNodeError_1 = require("./KeyNodeError");
exports.ROOT_KEYS = Symbol('Root Keys Lib');
const CHILDREN = Symbol('Child Keys Lib');
const DEPTH = Symbol('Depth Cache');
const PARENTS = Symbol('Parents Cache');
class BaseKeyNode extends String {
    constructor(key, parent) {
        super(key);
        this[_a] = new Map();
        const keyLiteral = this.toString();
        const parentIsRootKeyLib = parent instanceof Map;
        const siblingLib = parentIsRootKeyLib ? parent : parent[CHILDREN];
        if (siblingLib.has(keyLiteral)) {
            throw new KeyNodeError_1.default(`'${keyLiteral}' already exists in sibling set.` +
                `  Sibling key literals must be unique.`, siblingLib.get(keyLiteral));
        }
        siblingLib.set(keyLiteral, this);
        if (parentIsRootKeyLib) {
            this[exports.ROOT_KEYS] = siblingLib;
            this.PARENT = null;
            this.IS_ROOT_KEY = true;
        }
        else {
            this.PARENT = parent;
            this.IS_ROOT_KEY = false;
        }
    }
    //Accessors 
    get isTerminalKey() {
        return this[CHILDREN].size === 0;
    }
    get numChildren() {
        return this[CHILDREN].size;
    }
    get depth() {
        //Get and cache depth; lazy
        if (this[DEPTH] === undefined) {
            let depth = 0;
            for (const pKey of this.parents()) {
                depth++;
            }
            this[DEPTH] = depth;
        }
        return this[DEPTH];
    }
    get [(_a = CHILDREN, Symbol.toStringTag)]() {
        return this.constructor.name;
    }
    //Mehtods
    hasChild(childKey) {
        return this[CHILDREN].has(childKey);
    }
    getChild(childKey) {
        const children = this[CHILDREN];
        return children.has(childKey) ? children.get(childKey) : null;
    }
    hasSibling(siblingKey) {
        const sliblingKeyLib = this.IS_ROOT_KEY ? this[exports.ROOT_KEYS] : this.PARENT[CHILDREN];
        return sliblingKeyLib.has(siblingKey) ?
            sliblingKeyLib.get(siblingKey) !== this : false;
    }
    getSibling(siblingKey) {
        if (this.hasSibling(siblingKey)) {
            const sliblingKeyLib = this.IS_ROOT_KEY ? this[exports.ROOT_KEYS] : this.PARENT[CHILDREN];
            return sliblingKeyLib.get(siblingKey);
        }
        ;
        return null;
    }
    children() {
        //NOTE: Not cached b/c can increase.
        return this[CHILDREN].values();
    }
    parents() {
        //Get and cache parents; lazy
        if (this[PARENTS] === undefined) {
            let pKeys = new Set();
            let pKey = this.PARENT;
            while (pKey !== null) {
                pKeys.add(pKey);
                pKey = pKey.PARENT;
            }
            this[PARENTS] = pKeys;
        }
        return this[PARENTS].values();
    }
    *siblings() {
        //NOTE: Not cached b/c can increase.
        const siblingsIter = this.IS_ROOT_KEY ? this[exports.ROOT_KEYS].values() : this.PARENT.children();
        for (const sib of siblingsIter) {
            if (sib !== this) {
                yield sib;
            }
        }
    }
}
exports.BaseKeyNode = BaseKeyNode;
class KeyNode extends BaseKeyNode {
}
exports.KeyNode = KeyNode;
