// path-notation/build/PathNotation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const KEYS = Symbol();
/**
  @description Dot-notation based path string which yields key literals, allows for the inclusion of keys with "." characters, and provides node meta data and other useful utilities.
  @extends String
*/
class PathNotation extends String {
    /**
      @description Creates a PathNotation instance.
      @arg {string | number | String | Iterable<string | number | String>} pathNotationOrKey -
        Key in path notation e.g. "foo.bar['baz.qux']" or a single path key.
      @arg {...(string | number | String)} [keys] - Optional child path keys.
    */
    constructor(pathNotationOrKey, ...keys) {
        const keysToSet = [];
        switch (typeof pathNotationOrKey) {
            case 'string':
                keysToSet.push(...PathNotation.pathNotationToKeys(pathNotationOrKey));
                break;
            case 'number':
                keysToSet.push(...PathNotation.pathNotationToKeys(pathNotationOrKey.toString()));
                break;
            default:
                if (pathNotationOrKey instanceof String && !(pathNotationOrKey instanceof PathNotation)) {
                    keysToSet.push(pathNotationOrKey.toString());
                    break;
                }
                for (const keyLiteral of pathNotationOrKey) {
                    keysToSet.push(keyLiteral.toString());
                }
        }
        for (const keyLiteral of keys) {
            keysToSet.push(keyLiteral.toString());
        }
        super(PathNotation.keysToPathNotation(keysToSet));
        this[KEYS] = keysToSet;
    }
    //Accessors
    /**
      @description Number of key literals in path.
      @type {number}
    */
    get numKeys() {
        return this[KEYS].length;
    }
    /**
      @description Zero based node depth of terminal key in path.
      @type {number}
    */
    get depth() {
        return this[KEYS].length - 1;
    }
    /**
      @description First key literal in path.
      @type {string}
    */
    get rootKey() {
        return this[KEYS][0];
    }
    /**
      @description Last key literal in path.
      @type {string}
    */
    get terminalKey() {
        return this[KEYS][this.depth];
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    //Methods
    /**
      @description Yields key literals from path.
      @generator
      @yields {string} The next key litreral in path.
    */
    *keys() {
        let i = 0;
        for (let key of this[KEYS]) {
            i++;
            let remainingPath = yield key;
            while (remainingPath === true) {
                remainingPath = yield PathNotation.keysToPathNotation(this[KEYS].slice(i));
            }
        }
    }
    /**
      @description Returns parent path.
      @arg {number} keyDepth - 0 based index of parent path.  Cannot be greater than [PathNotation.depth]{@link Path#depth}.
    */
    parentPath(keyDepth) {
        const DEPTH = this.depth;
        if (keyDepth >= DEPTH) {
            throw new Error(`A parent 'PathNotation.depth' must be less than it's children 'PathNotation.depth(s)'.`);
        }
        return new PathNotation(this[KEYS].slice(0, keyDepth + 1));
    }
    *[Symbol.iterator]() {
        yield* this[KEYS];
    }
    /**
      @description Yileds key literals dot-notated path.
      @generator
      @yields {string} The next key litreral in path.
    */
    static *pathNotationToKeys(path) {
        const escapedKeys = new Map();
        //["@key"]
        path = path.replace(/\["(.+)"\]/g, (m, key) => {
            let placeHolder;
            do {
                placeHolder = `{%${Math.floor(Math.random() * 10 ** 4).toString()}%}`;
            } while (escapedKeys.has(placeHolder));
            escapedKeys.set(placeHolder, key);
            return `.${placeHolder}.`;
        });
        //['@key']
        path = path.replace(/\['(.+)'\]/g, (m, key) => {
            let placeHolder;
            do {
                placeHolder = `{%${Math.floor(Math.random() * 10 ** 4).toString()}%}`;
            } while (escapedKeys.has(placeHolder));
            escapedKeys.set(placeHolder, key);
            return `.${placeHolder}.`;
        });
        //Remove trailing '.'
        if (path.slice(-1) === '.') {
            path = path.slice(0, -1);
        }
        for (let key of path.split('.')) {
            yield escapedKeys.has(key) ? escapedKeys.get(key) : key;
        }
    }
    /**
      @description Returns dot-notated path.
      @returns {string} Dot-notated path.
    */
    static keysToPathNotation(keys) {
        let pathNotation = '';
        let lastWasBracketNotation = false;
        for (let key of keys) {
            if (key.indexOf('.') > -1) {
                pathNotation += `["${key}"]`;
                lastWasBracketNotation = true;
            }
            else if (lastWasBracketNotation) {
                pathNotation += `${key}`;
                lastWasBracketNotation = false;
            }
            else {
                pathNotation += `.${key}`;
            }
        }
        //Remove leading "."
        return pathNotation.slice(1);
    }
}
exports.default = PathNotation;
